tec: Update the lib_opml

marienfressinaud opened the pull request:
This is a pretty big PR which does mainly three things:

- it updates lib\_opml to its latest version (0.5.0 at this time, which will probably be the 1.0.0)
- it refactors the `ImportService` to apply two fixes I didn't accept in lib\_opml (cf. https://github.com/marienfressinaud/lib_opml/issues/3 and https://github.com/marienfressinaud/lib_opml/issues/5)
- the refactor also fixes an issue when counting the categories (existing categories were counted twice). The issue is still present for feeds, but I'm not sure what's the best strategy.

I didn't expect to refactor the code initially, but it happened that it was the easiest solution to fix the issues mentioned above. Also, I wasn't happy with the previous code architecture.

Overall, I think the new algorithm is a bit simpler to understand. Before: categories and feeds were created by recursion while we were browsing the OPML array. After: the OPML array is preprocessed by recursion, then categories and feeds are created sequentially.

lib\_opml 0.4 brings few interesting things for FreshRSS:

- it adds support for namespaces in a simpler way than the previous implementation
- it has a better support of the specification (e.g. special attributes and elements are parsed to their specified type)
- the non-strict mode accepts more violations of the specification
- it removes the dependency to simplexml
- all the code is tested, which means a more robust lib

On the other hand, the lib is a bit bigger than before.

Note that the repository moved to https://framagit.org/marienfressinaud/lib_opml.

How to test the feature manually:

1. with an existing user, add few feeds
2. make sure to have a HTML + XPath feed, and feed with an article CSS selector
3. export the OPML of the user
4. verify the custom FRSS options are exported as well
5. create an other user and import the OPML
6. verify the XPath options and the article CSS selector are set on the corresponding feeds

Pull request checklist:

- [x] clear commit messages
- [x] code manually tested
- [ ] unit tests written (optional if too hard)
- [x] documentation updated

[Additional information can be found in the documentation](https://github.com/FreshRSS/FreshRSS/tree/edge/docs/en/developers/04_Pull_requests.md).

marienfressinaud committed:
fix: List details of feeds for OPML exportation

The details are necessary to export the XPath information, the CSS full
content path and read actions filters.

marienfressinaud committed:
Update LibOpml to 0.4.0

marienfressinaud committed:
Refactor OPML importation to be more robust

First, it fixes two regressions introduced by the update of lib_opml:

- title attribute is used when text attribute is missing;
- the OPML category attribute is used as a fallback for feeds categories.

In a related way, if also fixes a problem when a feed had both a parent
category outline and a category attribute. Before, it only considered the
attribute as its category, but now it considers the parent outline.

Then, it counts category limit correctly by not increasing
`$nb_categories` if the category already exists.

marienfressinaud committed:
fix: Fix undefined GLOB_BRACE on Alpine

The manual states that:

> Note: The GLOB_BRACE flag is not available on some non GNU systems,
> like Solaris or Alpine Linux.

This generated an error on Alpine.

Reference: https://www.php.net/manual/function.glob.php

marienfressinaud committed:
Exclude lib_opml from the CodeSniffer

marienfressinaud committed:
Fix variable names when logging some errors

marienfressinaud committed:
Fix catch of LibOpml Exception

marienfressinaud committed:
Make sure to declare the category

marienfressinaud committed:
Exclude lib_opml from PHPStan analyze

marienfressinaud commented:
PHPStan is not able to discover the LibOpml class and functions, I'll fix that later.

Frenzie reviewed (commented):
Nothing looks out (or in other words, lgtm), but I didn't read it in great detail.

Alkarex commented on app/views/helpers/export/opml.phtml:
```diff
@@ -2,9 +2,12 @@
 
 $opml_array = array(
+	'namespaces' => array(
+		'frss' => FreshRSS_Export_Service::FRSS_NAMESPACE,
```
👍🏻

marienfressinaud committed:
Disable markdownlint for lib_opml

marienfressinaud committed:
Fix typos

marienfressinaud commented:
Finally, it works :) I disabled a bunch of checks as lib\_opml is an external library, but I fixed upstream most of the errors I saw (will be in the next release of lib\_opml).

The biggest change I've done is to the PHPStan configuration file: I think the `scanDirectories` wasn't working correctly because the folders there were also in the `excludePaths`. I've adapted the `excludePaths` directive: I think it works better now.

Alkarex commented on lib/lib_opml/src/LibOpml/LibOpml.php:
```diff
@@ -0,0 +1,682 @@
+    /**
+     * The list of valid head elements.
+     */
+    public const HEAD_ELEMENTS = [
```
With PHP 7.0:
```
PHP Parse error:  syntax error, unexpected 'const' (T_CONST), expecting variable (T_VARIABLE) in ./lib/marienfressinaud/lib_opml/src/LibOpml/LibOpml.php on line 67
```

I am not sure of the added value of `public` here but if it stays, this PR will have to wait till we drop PHP 7.0, which is soon: https://github.com/FreshRSS/FreshRSS/milestones

1.20.x will be the last supporting PHP 7.0 with PHP 7.2+ required from 1.21.0

marienfressinaud replied:
The linter asks me to add visibility to the constants. I'm a bit reluctant to change the default configuration so I guess we'll wait for the 1.21 if you don't mind :sweat_smile: I commit myself to update the PR if there are conflicts with the future changes :)

marienfressinaud replied:
Btw, I realize I officially only support PHP 7.4+ with lib\_opml, I'll see what I can do to support 7.2+

marienfressinaud replied:
PHP 7.2+ is now officially supported by lib\_opml: https://framagit.org/marienfressinaud/lib_opml/-/commit/a8dfa46a80202f3b5e71555e05eb7a4125d74b5b

Alkarex replied:
Ready for PHP 7.2+ https://github.com/FreshRSS/FreshRSS/pull/4848

Alkarex committed:
Use auto-loading and allow updates via Composer

Alkarex commented:
@marienfressinaud It looks like your PR does not allow contributions (*Allow edits by maintainers*), so I had to publish another branch https://github.com/flusio/FreshRSS/pull/1 for my changes:

https://github.com/Alkarex/FreshRSS/commit/49365963e2b3a452e55ff6712a5f9fef341c6e26

Please double-check: I have added auto-loading instead of manual requires, and allowed optional updates via Composer on the model of what we did for https://github.com/FreshRSS/FreshRSS/pull/4329 and https://github.com/FreshRSS/FreshRSS/pull/4368

Please be aware that your Composer package is full of irrelevant files, in particular the `.git/` folder ;-)

Alkarex committed:
Fix broken links to lib_opml

marienfressinaud commented:
> It looks like your PR does not allow contributions (Allow edits by maintainers), so I had to publish another branch https://github.com/flusio/FreshRSS/pull/1 for my changes:

I don't know where I can enable contributions :thinking:

Alkarex commented:
> I don't know where I can enable contributions 🤔

When you create the PR, there is an option, otherwise for an existing PR:

![image](https://user-images.githubusercontent.com/1008324/173930033-14328eb5-8aa6-4637-ab77-359b3bf51555.png)

Alkarex commented:
P.S. But if you accept + merge my PR, it should do the trick

marienfressinaud commented:
> P.S. But if you accept + merge my PR, it should do the trick

I hope so because I don't have this option (maybe because I'm a maintainer myself?)

Alkarex commented:
It worked :-)

> I hope so because I don't have this option (maybe because I'm a maintainer myself?)

I do not think that is the reason. This is a feature used all the time, also between multiple maintainers / owners.

Alkarex commented:
Side note: talking about OPML, I am just done with this new feature with dynamic OPML / reading lists (both as client and as server) https://github.com/FreshRSS/FreshRSS/pull/4407

marienfressinaud commented:
> I do not think that is the reason. This is a feature used all the time, also between multiple maintainers / owners.

Still mysterious to me, I don't remember seeing the option when I opened the PR, and it was definitely not there after :thinking: 

> Side note: talking about OPML, I am just done with this new feature with dynamic OPML / reading lists (both as client and as server) https://github.com/FreshRSS/FreshRSS/pull/4407

And that's an awesome feature in my opinion! I'm very happy to see it coming to FRSS :) I don't have much time to review the PR, but don't hesitate to ping me if you have specific questions (I will not be able to answer before next week though)

Alkarex commented:
> > Side note: talking about OPML, I am just done with this new feature with dynamic OPML / reading lists (both as client and as server) #4407
> 
> And that's an awesome feature in my opinion! I'm very happy to see it coming to FRSS :) I don't have much time to review the PR, but don't hesitate to ping me if you have specific questions (I will not be able to answer before next week though)

No questions at the moment, but looking for some cools examples in both directions (client and/or server) :-)
Worth noting that it can also be combined with the new XPath Web scraping feature (to share the XPath settings via OPML and/or the results as RSS).

Alkarex commented:
We now have moved to PHP 7.2+ so this PR can be updated @marienfressinaud

marienfressinaud committed:
Merge branch 'edge' into tec/update-libopml

marienfressinaud committed:
Bring back the ability to import the OPML frss:opmlUrl attribute

marienfressinaud committed:
Refactor the logs of OPML errors

marienfressinaud committed:
Update lib_opml to the version 0.5.0

marienfressinaud commented:
It should be good now! It was quite complicated because the code changed a lot since I opened this PR (especially because of https://github.com/FreshRSS/FreshRSS/pull/4407), but I think I managed to include all the changes. I changed a bit the signature of `ImportService->importOpml`:

- I merged the `$parent_cat` and the `$flatten` parameters in one `$forced_category`
- it no longer returns the list of categories (they weren't used, even during the dynamic OPML refresh)

I also updated lib\_opml to the 0.5.0, which should become the 1.0.0 once it has been tested in FreshRSS for a while.

chuxubank commented:
Looking forward for this PR get merged!

I'm curious about whether this PR enable the ability of dynamic import opml creating categories?

Alkarex commented:
> I'm curious about whether this PR enable the ability of dynamic import opml creating categories?

No change in features with this PR. The dynamic OPML stuff was done in https://github.com/FreshRSS/FreshRSS/pull/4407

Alwaysin commented:
This (removal of lib_opml.php) broke "Import YouTube Takeout" plugin: https://github.com/simsor/freshrss-extensions/issues/1
